File: android/app/src/main/java/com/bluelink/android/files/ChatThumbnailGeometry.kt
package com.bluelink.android.files

/** Natural-size small images; fit until the short edge is 48, then center-crop the long edge. */
internal data class ChatThumbnailGeometry(val width: Double, val height: Double, val scale: Double,
    val cropX: Double, val cropY: Double, val cropWidth: Double, val cropHeight: Double) {
    val imageWidth get() = cropWidth * scale
    val imageHeight get() = cropHeight * scale
    val insetX get() = (width - imageWidth) / 2
    val insetY get() = (height - imageHeight) / 2

    companion object {
        fun calculate(width: Double, height: Double, maxWidth: Double = 190.0, maxHeight: Double = 126.0): ChatThumbnailGeometry {
            val minimum = 48.0
            require(width.isFinite() && height.isFinite() && width > 0 && height > 0 &&
                maxWidth.isFinite() && maxHeight.isFinite() && maxWidth >= minimum && maxHeight >= minimum)
            val shortEdge = minOf(width, height)
            val fit = minOf(1.0, maxWidth / width, maxHeight / height)
            val scale = if (shortEdge <= minimum) 1.0 else maxOf(fit, minimum / shortEdge)
            val viewportWidth = (width * scale).coerceIn(minimum, maxWidth)
            val viewportHeight = (height * scale).coerceIn(minimum, maxHeight)
            val cropWidth = minOf(width, viewportWidth / scale)
            val cropHeight = minOf(height, viewportHeight / scale)
            return ChatThumbnailGeometry(viewportWidth, viewportHeight, scale, (width - cropWidth) / 2,
                (height - cropHeight) / 2, cropWidth, cropHeight)
        }
    }
}

File: windows/BlueLink.App/Files/ChatThumbnailGeometry.cs
namespace BlueLink.Files;

/// <summary>Fit until the short edge reaches 48, then center-crop instead of shrinking further.</summary>
internal sealed record ChatThumbnailGeometry(double Width, double Height, double Scale,
    double CropX, double CropY, double CropWidth, double CropHeight)
{
    public double ImageWidth => CropWidth * Scale;
    public double ImageHeight => CropHeight * Scale;
    public double InsetX => (Width - ImageWidth) / 2;
    public double InsetY => (Height - ImageHeight) / 2;

    public static ChatThumbnailGeometry Calculate(double width, double height, double maxWidth = 240, double maxHeight = 144, double minimum = 48)
    {
        if (!double.IsFinite(minimum) || minimum <= 0 || !double.IsFinite(width) || !double.IsFinite(height) || width <= 0 || height <= 0 ||
            !double.IsFinite(maxWidth) || !double.IsFinite(maxHeight) || maxWidth < minimum || maxHeight < minimum)
            throw new ArgumentOutOfRangeException(nameof(width));
        var shortEdge = Math.Min(width, height);
        var fit = Math.Min(1, Math.Min(maxWidth / width, maxHeight / height));
        var scale = shortEdge <= minimum ? 1 : Math.Max(fit, minimum / shortEdge);
        var viewportWidth = Math.Clamp(width * scale, minimum, maxWidth);
        var viewportHeight = Math.Clamp(height * scale, minimum, maxHeight);
        var cropWidth = Math.Min(width, viewportWidth / scale);
        var cropHeight = Math.Min(height, viewportHeight / scale);
        return new(viewportWidth, viewportHeight, scale, (width - cropWidth) / 2, (height - cropHeight) / 2, cropWidth, cropHeight);
    }
}
